strawberryfields/api/result.py

```diff
@@ -107,6 +102,28 @@ def state(self):
             raise AttributeError("The state is undefined for a stateless computation.")
         return self._state
 
+    @staticmethod
+    def combine_samples(samples_list, mode_order):
+        """Combine samples measured at different times into one nested array
+
+        Args:
+            samples_list (list): the sample measurements in a list
+            mode_order (list): the mode order for the measured modes
+
+        Returns:
+            array: the samples in raising mode order with shape ``(shots, measured_modes)``
+        """
+        ret = [[] for _ in samples_list[0]]
+        for i, samples in enumerate(samples_list):
+            for j, single_sample in enumerate(samples):
+                for s in single_sample:
+                    ret[j].append(s)
+
+        for i, sam in enumerate(ret):
+            ret[i] = [j for _, j in sorted(zip(mode_order, sam))]
+
+        return np.array(ret)
```

Comment:
Although, you would have to be smart in the `dtype` used when initializing `ret`; either an int, float, or complex depending on what types of measurements are present.

Reply:
Just had another look at it, and there's an issue with the `for m, s in zip(mode_order, samples_list)` since `mode_order` only would include the measured modes, which could include values larger than the number of measured modes `modes`. E.g. if `mode_order = [array([1, 3, 4]), array([2, 6])]` using 7 modes but only measuring 5. Then, `modes = 5` and thus `ret[:, m]` would cause an out-of-bounds error for `m=array([2, 6])`.

Not sure if it's worth having another part re-calculating `mode_order` into, in the example above, `mode_order = [array([0, 2, 3]), array([1, 4])]` is worth it, either in `BaseEngine._run` or 
in `combine_samples`. It could be something like:

```python
sort_order = [
    np.array([r.ind for r in c.reg])
    for c in p.circuit
    if "Measure" in c.op.__str__()
]

s = [tup[0] for tup in sorted(enumerate(sort_order), key=lambda x: x[1])]

res = []
start = 0
for x in sort_order:
    res.append(np.array(s[start:start + len(x)]))
    start = len(x)
sort_order = res
```

but that's quite a lot of _extra_ code. What do you think about this, or do you have another idea for a solution? There's probably a way to to this in a nicer way, but I cannot really think of any solution right now. :sleepy:

Reply:
this might be quite a complicated task indeed and would need more refined logic I guess! Having a nested `np.argsort` would be cool if the list was flattened, but flattening and unflattering it might also be costly.

I cooked up this one (on first glance I think there's no clear advantage to it at all :smile: just thought of leaving it here if it gives some ideas):
```python
import numpy as np

mode_order = [np.array([1, 3, 4]), np.array([2, 6])]
modes = np.concatenate(mode_order)
all_modes = np.arange(np.max(modes))
remainders = set(all_modes) - set(modes)

for modes in mode_order:
    for i in remainders:
        modes[modes>=i] -= 1
```
```
[array([0, 2, 3]), array([1, 4])]
```

Reply:
> if mode_order = [array([1, 3, 4]), array([2, 6])] using 7 modes but only measuring 5. Then, modes = 5 and thus ret[:, m] would cause an out-of-bounds error for m=array([2, 6])

Oh! Good catch. This sounds like a solution for `np.argsort` and `np.split`, as @antalszava suggested, but then you have to also take into account the flattening.

The cost might not be that high, since we know the maximum level of nesting. E.g.,

```python
>>> mode_order = [np.array([0, 5, 2, 9, 6]),  np.array([3, 10])]
>>> shapes = np.cumsum([len(m) for m in mode_order])
array([5, 7])
>>> np.split(np.argsort(np.argsort(np.hstack(mode_order))), shapes)[:-1]
[array([0, 3, 1, 5, 4]), array([2, 6])]
```